### PR TITLE
[llms] Add Llama-3.1-8B Q4NX prefill + decode on NPU2 (split weight BOs past 4 GiB)

### PR DESCRIPTION
**Pull request by erwei-xilinx**
Reproduces FastFlowLM's Llama-3.1-8B on NPU2, prefill and decode, as
`programming_examples/llms/llama31_8b_q4nx`.

The whole decoder layer runs on the AIE array — RMSNorm, QKV, RoPE, flash
attention over a device-resident KV cache, O, FFN — plus the LM head, in one
dispatch per token for 32 layers. No CPU attention, no host KV cache. Weights
are FastFlowLM's 4-bit `model.q4nx` bundle.

The attention topology is identical to Llama-3.2-1B/3B (`ATTN_IMPL_2x4x1`, 8 kv
heads, DH=128), so the per-CU KV geometry carries over unchanged and only the
proj/FFN widths grow. What is new is all a consequence of K=4096 and 8B of
weights.

## The weights do not fit one buffer

A shim BD's byte offset is a `uint32` end to end, so a single BO is addressable
over at most 4 GiB. This model's 32 layer slabs plus the lm-head are
**4.375 GiB**: the offsets wrapped and every logit came back NaN.

`DECODE_WGROUP=8` (the knob #1836 added for Qwen3-8B) splits the weights into
four 1.02 GiB layer groups plus a 0.31 GiB lm-head buffer, each addressed from
its own base. The host slices its BOs to match, reading the geometry off the
engine rather than the environment so the two cannot disagree.

Getting there took a while because every data-level suspect was innocent: the
weight cache matched byte for byte (its 0.225% NaN fraction is 4-bit payload
aliased as bf16, identical to the known-good 3B cache), the stack trim was
exonerated by rebuilding a working 3B at the same value, and the host inputs
were all clean. The weights were fine; the addressing wrapped.

## Other K=4096 consequences

- **`DECODE_STACK=8064`.** The rms core's seven K-wide L1 activation buffers
  leave under 8 KiB, so the 10240 default fails buffer allocation at build time.
- **Two FFN GEMM registry shapes** at intermediate=14336 (the attention shapes
  came free from Qwen3-8B, which shares K=4096). Worth flagging for whoever
  reads the registry next: the Qwen3-8B N=12288 entry concludes fused-cast is
  legal only at `tile_k_l2=64` because the f32-out B-tile stride must stay under
  1048576. That is **not** what binds at N=14336 — `tile_k_l2=128` (stride
  1835008) places, runs, and is 8% faster at identical accuracy. `tile_k_l2=256`
  is left uncharacterized and the note says so: aiecc ran over an hour without
  finishing, so a faster point may exist above 128.
- **LM-head GEMV tiling.** Its L2 A stage overflows by exactly 128 B at the
  default 8x8; same `tile_m` halving the Qwen3-8B sibling already carries.

## Untied LM head

Unlike 1B/3B, this model sets `tie_word_embeddings=false` and the bundle ships a
real quantized `lm_head`. The shared requant path hard-coded the tied case, so it
takes a `tie_lm_head` flag. Measured `cosine(bundle lm_head, embed_tokens) =
0.017` — essentially orthogonal, so the tied default would have produced a
silently wrong model rather than a build error.

## The Paris gate was failing a correct build

The prefill's weight-integrity smoke was seeded from the 1B verbatim, and its
prompt does not discriminate for this model. On "The capital of France is",
Llama-3.1-8B-Instruct ranks `" a"` at 16.000 over `" Paris"` at 15.938 — and the
HF bf16 reference ranks them the same way. What identified the gate rather than
the model was that the NPU prefill's top-4 matched HF's in order. Regated on
"The capital city of France is called", where `" Paris"` leads by 3.4 logits.

## Measured on NPU2

| gate | result |
|---|---|
| `make verify` (2 prompts x 32 tokens, k=5, vs HF bf16) | 2 / 2 PASS |
| `make verify-full` (8 prompts) | 8 / 8 PASS |
| `make verify-paris` (prefill-only smoke) | PASS (argmax 12366) |
| `make profile N_TOKENS=128` | TTFT 1.61 s, decode 10.7 tok/s |
| `make profile-prefill` (CTX=2048) | 6.22 s, 330 tok/s, 96% NPU |

At hidden=14336 the fused O+FFN dominates prefill (3900 ms over 32 layers)
ahead of attention (1452 ms), the reverse of the Qwen3-8B sibling. FastFlowLM
publishes no figure for this model that I could source, so unlike Qwen3-8B the
README quotes no reproduction ratio.

All of the above was re-measured after rebasing onto current main, on a
toolchain rebuilt to include #1826 (`boundShimFeedBursts`), since that pass emits
the shim feed BDs the split weight BOs depend on. It is inert here, as its own
comment predicts — the timings are identical within noise.

## Consistency

Audited against `llama32_1b_q4nx`, `llama32_3b_q4nx`, `gemma3_4b_q4nx`,
`phi4_mini_q4nx`, `qwen3_8b_q4nx` and `qwen25_3b_q4`. File inventory, Makefile
target set and header, `verify_adapter.py` surface, `.gitignore`, the three
registration points and the nightly/website wiring all already matched; the
`hfweights_*` REQUIRES features and the `extract_perf.py` parse were verified
against the real profile output rather than assumed. Two things were fixed: the
compile lits had the prefill/decode convention inverted relative to the four
siblings that share this example's target set, and the README was missing five
canonical sections — including the decode staircase, which this example ships
but never documented.

### CODE REVIEW

File: programming_examples/llms/llama31_8b_q4nx/q4nx_decode_8b.py [hunk 1]
```diff
@@ -0,0 +1,426 @@
...
+        RMS_in = list(z["RMS_in"].view(bfloat16))
+        RMS_post = list(z["RMS_post"].view(bfloat16))
+        qm = Q4nxModel(model_source)
+        # Only the input embedding + final norm are read here; the LM head lives in
+        # the requant cache above (untied, from the bundle's own lm_head tensor), so
+        # embed_norm_lmhead()'s tied third return is deliberately not used.
+        embed = qm.bf16("model.embed_tokens.weight")
+        final_norm = qm.bf16("model.norm.weight")
+        self.embed = np.asarray(embed, np.float32).reshape(self.VOCAB_SIZE, self.K)
+        final_norm = np.asarray(final_norm, bfloat16)
```
[the remark this answers is not in the document]
Comment (erwei-xilinx) [reply programming_examples/llms/llama31_8b_q4nx/q4nx_decode_8b.py:237]: Fixed in 189e3ed8f.
